### PR TITLE
[core] Remove use variable length array

### PR DESCRIPTION
**Pull request by quink-black**

### CODE REVIEW

**Comment by maxsharabayko** [comment]
I would propose using the `FixedArray` then ([from utilities.h](https://github.com/Haivision/srt/blob/master/srtcore/utilities.h#L418)).
It can't (and is not expected to) be resized, unlike `std::vector`. Even though in this case of a local variable it is not that important.

**Comment by quink-black** [comment]
> I would propose using the `FixedArray` then ([from utilities.h](https://github.com/Haivision/srt/blob/master/srtcore/utilities.h#L418)). It can't (and is not expected to) be resized, unlike `std::vector`. Even though in this case of a local variable it is not that important.

Could you give some hints about the benefits of `FixedArray` other than forbidden of resize? There is no reallocation here, so I guess the overhead of `std::vector` can be negligible.

**Comment by maxsharabayko** [comment]
> Could you give some hints about the benefits of `FixedArray` other than forbidden of resize? There is no reallocation here, so I guess the overhead of `std::vector` can be negligible.

It is a perfectly fair question, but I have no unambiguously correct answer to it. I will give my thoughts, while I do not insist on using the `FixedArray`. Especially given we talk about a local variable with a short lifetime.

In this very place, we want to have an array of `epoll_event`. The ideal would be to use `std::array`, except we don't know the size at compile time. Therefore we need an array allocated in run-time, preferably destructing itself properly, so we don't need to call `delete []`.
Based on this description a conventional guideline is to use `std::vector`.

However, `std::vector` is much more powerful than just a dynamically allocated array. Thus we get two disadvantages:
1. If a non-const `std::vector` is used, we don't have a guarantee that somewhere it will not be resized. The use of `vector` itself does not provide this intent to the reader of the code. In this case of a local scope - not a big deal.
2. Slightly extra footprint. We need a pointer and array size, we get a bit more. E.g. see the below implementation of the `std::vector` having two-pointers and an additional capacity pointer. (_The `FixedArray` currently also has an extra pointer pointing to an error string I brought in PR #2184. A quick and dirty fix to be re-worked._)

Thus in this "proposal" my main motivation is the first item from the list above: to show intent it is just an array we don't plan to resize.

```c++
template <class _Tp, class _Allocator>
class __vector_base
    : protected __vector_base_common<true>
{
    // ...
    pointer                                         __begin_;
    pointer                                         __end_;
    __compressed_pair<pointer, allocator_type> __end_cap_;
    // ...
}

```

**Comment by maxsharabayko** [comment]
PR #2298 adds `srt::FixedArray::data()` that (once merged) can be used here instead of `&ev[0]`.

File: srtcore/epoll.cpp [hunk 1]
```diff
@@ -637,8 +638,8 @@ int CEPoll::wait(const int eid, set<SRTSOCKET>* readfds, set<SRTSOCKET>* writefd
 #ifdef LINUX
                 const int max_events = ed.m_sLocals.size();
                 SRT_ASSERT(max_events > 0);
-                epoll_event ev[max_events];
-                int nfds = ::epoll_wait(ed.m_iLocalID, ev, max_events, 0);
+                srt::FixedArray<epoll_event> ev(max_events);
+                int nfds = ::epoll_wait(ed.m_iLocalID, &ev[0], max_events, 0);
```
Comment (maxsharabayko) [inline srtcore/epoll.cpp:642]:
```suggestion
                int nfds = ::epoll_wait(ed.m_iLocalID, ev.data(), ev.size(), 0);
```

File: srtcore/epoll.cpp [hunk 2]
```diff
@@ -660,9 +661,9 @@ int CEPoll::wait(const int eid, set<SRTSOCKET>* readfds, set<SRTSOCKET>* writefd
                 struct timespec tmout = {0, 0};
                 const int max_events = ed.m_sLocals.size();
                 SRT_ASSERT(max_events > 0);
-                struct kevent ke[max_events];
+                srt::FixedArray<struct kevent> ke(max_events);
 
-                int nfds = kevent(ed.m_iLocalID, NULL, 0, ke, max_events, &tmout);
+                int nfds = kevent(ed.m_iLocalID, NULL, 0, &ke[0], max_events, &tmout);
```
Comment (maxsharabayko) [inline srtcore/epoll.cpp:666]:
```suggestion
                int nfds = kevent(ed.m_iLocalID, NULL, 0, ke.data(), ke.size(), &tmout);
```

**Review by maxsharabayko — CHANGES_REQUESTED** [review]

**Pull request merged**